.taskcluster.yml

```diff
@@ -1,39 +1,60 @@
-      maxRunTime: 3600
-      image: taskcluster/upload-project-docs:latest
-      features:
-        taskclusterProxy:
-          true
-      command:
-        - /bin/bash
-        - '--login'
-        - '-cx'
-        - >-
-          git clone {{event.head.repo.url}} repo &&
-          cd repo &&
-          git config advice.detachedHead false &&
-          git checkout {{event.head.sha}} &&
-          export DOCS_PROJECT=docker-exec-websocket-server DOCS_TIER=libraries DOCS_FOLDER=docs DOCS_README=README.md &&
-          upload-project-docs
-    metadata:
-      name: "docker-exec-websocket-server docs upload"
-      description: "Upload documentation for this project"
-      owner: '{{ event.head.user.email }}'
-      source: '{{ event.head.repo.url }}'
+  $let:
+    proj_name: "docker-exec-websocket-server"
+    head_rev:
+      $if: tasks_for == "github-pull-request"
+      then: ${event.pull_request.head.sha}
+      else: ${event.after}
+    repository:
+      $if: tasks_for == "github-pull-request"
+      then: ${event.pull_request.head.repo.html_url}
+      else: ${event.repository.html_url}
+    environments:
+      - node_image_tag: 16-bullseye
+        description: Run tests with node v16
+      - node_image_tag: 14-bullseye
+        description: Run tests with node v14
+      - node_image_tag: 12-bullseye
+        description: Run tests with node v12
+      - node_image_tag: "10"
+        description: Run tests with node v10 (unsupported)
+      - node_image_tag: current-bullseye
+        description: Run tests with current node version
+  in:
+    $if: tasks_for == "github-pull-request" && event["action"] in ["opened","reopened","synchronize"]
+    then:
+      $map: {$eval: environments}
+      each(env):
+        taskId: {$eval: as_slugid(proj_name + ":node:" + env.node_image_tag + ":tests")}
+        deadline:
+          $fromNow: 1 day
+        taskQueueId: proj-taskcluster/gw-ci-ubuntu-18-04
+        metadata:
+          name: node:${env.node_image_tag} tests
+          description: ${env.description}
+          owner: ${event.sender.login}@users.noreply.github.com
+          source: ${event.repository.url}
+        payload:
+          maxRunTime: 3600
+          env:
+            BUILDKIT_PROGRESS: plain  # auto / tty is over-verbose
+            REPO_DIR: {$eval: as_slugid(proj_name + ":node:" + env.node_image_tag + ":tests")}
+          command:
+            -
+              - /bin/bash
+              - -vxec
+              - >
+                git --version &&
+                git clone --no-progress "${repository}" "$REPO_DIR" &&
+                chmod a+wrx "$REPO_DIR" &&
+                cd "$REPO_DIR" &&
+                git config advice.detachedHead "false" &&
+                git checkout --no-progress "${head_rev}" &&
+                mkdir -p ~/bin &&
+                curl -sL "https://github.com/docker/compose/releases/download/1.29.2/docker-compose-$(uname -s)-$(uname -m)" -o ~/bin/docker-compose &&
```

Comment:
Is it worth having retries here to protect against network blips?

One option to do that would be to use the [mounts](https://github.com/taskcluster/taskcluster/blob/66941613e4242d69dd2aff3fb560359eb633d59c/workers/generic-worker/schemas/multiuser_posix.yml#L216-L225) feature, e.g. have the following inside the task `payload` (i.e. _sibling_ of `maxruntime`, `env`, `command`):

```
          mounts:
            - file: bin/docker-compose
              content:
                url: https://github.com/docker/compose/releases/download/1.29.2/docker-compose-Linux-x86_64
                sha256: f3f10cf3dbb8107e9ba2ea5f23c1d2159ff7321d16f0a23051d68d8e2547b323
```

Note, the `sha256` is optional, but I see that it is provided in the release, so might be a nice extra to have to protect against man-in-the-middle modifications.